core/src/main/java/io/camunda/migrator/Pagination.java

```diff
@@ -88,25 +96,58 @@ public List<T> toList() {
+      String activityInstanceId = variable.getActivityInstanceId();
+      Map<String, Object> variableMap = result.getOrDefault(activityInstanceId, new HashMap<>());
+      variableMap.putAll(variableResult);
+      result.put(activityInstanceId, variableMap);
+    });
+    return result;
+  }
+
+
+  public Map<String, Object> toVariableMapSingleActivity() {
+    Map<String, Object> variableResult = new HashMap<>();
+    processVariables((variable, typedValue) -> variableResult.putAll(getVariableResult(variable, typedValue)));
+    return variableResult;
+  }
+
   /**
    * Heads-up: this implementation needs to be null safe for the variable value.
    * Using streams might lead to undesired {@link NullPointerException}s.
    */
-  public Map<String, Object> toVariableMap() {
-    Map<String, Object> result = new HashMap<>();
+  private void processVariables(BiConsumer<VariableInstanceEntity, TypedValue> consumer) {
+    List<VariableInterceptor> interceptors = (List<VariableInterceptor>) context.getBeansOfType(
+        VariableInterceptor.class).values().stream().toList();
     toList().forEach(e -> {
-      VariableInstanceEntity var = (VariableInstanceEntity) e;
-      TypedValue typedValue = var.getTypedValue(false);
-      if (typedValue.getType().equals(ValueType.OBJECT)) {
-        // skip the value deserialization
-        result.put(var.getName(), typedValue.getValue());
-      } else if (typedValue.getType().equals(SpinValueType.JSON) || typedValue.getType().equals(SpinValueType.XML)) {
-        // For Spin JSON/XML, explicitly set the string value
-        result.put(var.getName(), typedValue.getValue().toString());
-      } else {
-        result.put(var.getName(), var.getValue());
+      VariableInvocation variableInvocation = new VariableInvocation((VariableInstanceEntity) e);
+      if (interceptors != null && !interceptors.isEmpty()) {
+        interceptors.forEach(i -> {
+          try {
+            i.execute(variableInvocation);
+          } catch (Exception ex) {
+            throw new VariableInterceptorException("An error occurred during variable transformation.", ex);
+          }
+        });
       }
+      VariableInstanceEntity variable = variableInvocation.getVariable();
+      TypedValue typedValue = variable.getTypedValue(false);
+      consumer.accept(variable, typedValue);
     });
-    return result;
+  }
+
+  private Map<String, Object> getVariableResult(VariableInstanceEntity variable, TypedValue typedValue) {
+    Map<String, Object> variableResult = new HashMap<>();
+    if (typedValue.getType().equals(ValueType.OBJECT)) {
+      variableResult.put(variable.getName(), typedValue.getValue());
+    } else if (typedValue.getType().equals(SpinValueType.JSON) || typedValue.getType().equals(SpinValueType.XML)) {
+      variableResult.put(variable.getName(), typedValue.getValue().toString());
+    } else {
+      variableResult.put(variable.getName(), variable.getValue());
+    }
+    return variableResult;
```

Comment:
The idea here is to do this in a default interceptor with order(0) and the rest will be executed afterwards. The last in line takes precedence. 
WIP: https://github.com/camunda/c7-data-migrator/pull/100/commits/4be96431ef8ad324534711197c48d1ec786f1dd4
I need to do more validation to confirm the approach.